grpc/nodejs/rules.bzl

```diff
@@ -14,116 +14,96 @@
 # You should have received a copy of the GNU Affero General Public License
 # along with this program.  If not, see <https://www.gnu.org/licenses/>.
 #
+load("@bazel_skylib//lib:paths.bzl", "paths")
+load("@rules_proto//proto:defs.bzl", "ProtoInfo")
 
-def _proto_sources(ctx):
-    inputs = []
-    for dep in ctx.attr.deps:
-        for src in dep[ProtoInfo].direct_sources:
-            inputs.append(src)
-    return DefaultInfo(files = depset(inputs))
+def _as_path(path, is_windows_host):
+    if is_windows_host:
+        return path.replace("/", "\\")
+    else:
+        return path
 
-
-def _unpack_proto_archive(ctx):
+def _get_outputs(target, ctx):
     outputs = []
-    for dep in ctx.attr.deps:
-        for src in dep[ProtoInfo].direct_sources:
-            outputs.extend([
-                ctx.actions.declare_file(
-                    src.path.replace('.proto', '_pb.d.ts')
-                ),
-                ctx.actions.declare_file(
-                    src.path.replace('.proto', '_pb.js')
-                ),
-            ])
+    root = target[ProtoInfo].proto_source_root
+    for source in target[ProtoInfo].direct_sources:
+        # test.proto -> {designated output dir}/test.ts
+        name = source.path.replace(source.extension, "ts")
+        dest = root
+        # if the dest is pwd then make sure that we do not strip subdirectories.
+        if dest == ".":
+            dest = source.dirname[len(ctx.label.package) + 1:]
+            if not dest:
+                dest = "."
+        output = ctx.actions.declare_file("/".join([dest, name])) # preserve dir structure for now
+        outputs.append(output)
+    return outputs
 
-    for dep in ctx.attr.grpc_deps:
-        for src in dep[ProtoInfo].direct_sources:
-            outputs.extend([
-                ctx.actions.declare_file(
-                    src.path.replace('.proto', '_grpc_pb.js')
-                ),
-                ctx.actions.declare_file(
-                    src.path.replace('.proto', '_grpc_pb.d.ts')
-                ),
-            ])
-    ctx.actions.run_shell(
-        inputs = [ctx.file.archive],
-        outputs = outputs,
-        mnemonic='x',
-        command = "mkdir -p {} && tar -xvf {} -C {}/{}".format(
-            ctx.label.package, ctx.file.archive.path,
-            ctx.var["BINDIR"], ctx.label.package
-        )
-    )
-    return DefaultInfo(files = depset(outputs))
+def _imported_protos(target, provided_sources = []):
+    proto_info = target[ProtoInfo]
+    source_root = proto_info.proto_source_root
+    if source_root == ".":
+        return [src.path for src in proto_info.direct_sources]
 
+    offset = len(source_root) + 1  # + '/'.
+    return [src.path[offset:] for src in proto_info.direct_sources]
 
-proto_sources = rule(
-    attrs = {
-        "deps": attr.label_list(
-            providers = [ProtoInfo],
+def _node_protoc(ctx):
```

Comment:
This implementation is derived from `protoc-gen-ts`, but with the paths fixed/adjusted for our use. Ideally we could remove this and just use `protoc-gen-ts`'s bazel rules, once they generate the expected outputs or use their rule in a compatible way.